File: src/python/client_node.py
#!/home/zanfranceschi/Projects/distributed-search/src/python/virtenv/bin/python
# -*- coding: utf-8 -*-

import json
import zmq
import uuid

ctx = zmq.Context()
socket_pub = ctx.socket(zmq.PUB)
socket_ack = ctx.socket(zmq.ROUTER)
socket_reply = ctx.socket(zmq.ROUTER)

socket_pub.bind("tcp://*:5000")
socket_ack.bind("tcp://*:5001")
socket_reply.bind("tcp://*:5002")

try:
	while True:
		q = raw_input("enter your search term: ")
		categories = raw_input("enter the categories you'd like to search (comma separated): ").split(',')

		search = { "q" : q, "categories" : categories }

		search_id = str(uuid.uuid4())

		print "id:", search_id

		for i in range(20):
			socket_pub.send(b'bosta de array de bytes')

		socket_pub.send("search_notification", zmq.SNDMORE) 		# topic
		socket_pub.send(search_id, zmq.SNDMORE) 					# searchId
		socket_pub.send("tcp://localhost:5001", zmq.SNDMORE) 		# ack endpoint
		socket_pub.send("tcp://localhost:5002", zmq.SNDMORE) 		# reply endpoint
		socket_pub.send("UTF-8", zmq.SNDMORE) 						# reply endpoint
		socket_pub.send(json.dumps(search)) 						# search object

		acks = []

		while True:
			if socket_ack.poll(100, zmq.POLLIN):
				msg_ack = socket_ack.recv_multipart()
				acks.append(msg_ack)
			else:
				break

		print len(acks), "acks received"

		results = []

		print "waiting results..."

		while len(results) < len(acks):
			if socket_reply.poll(15 * 1000):
				reply_message = socket_reply.recv_multipart()
				reply_search_id = reply_message[1]

				if reply_search_id == search_id:
					results.append(reply_message)
			else:
				break
		
		print len(results), "nodes responded"

		search_results = [json.loads(result[3]) for result in results]

		for search_result in search_results:
			print "node", search_result["searchNode"], "has", len(search_result["resultItems"]), "results"

		for search_result in search_results:
			result_items = search_result["resultItems"]
			for result_item in result_items:
				print result_item
except KeyboardInterrupt:
	socket_pub.close()
	socket_ack.close()
	socket_reply.close()
	ctx.term()
	print "bye"

File: src/java/distributed-search_client-node/src/org/distributedsearch/ClientNode.java
package org.distributedsearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.UUID;

import org.zeromq.ZFrame;
import org.zeromq.ZLoop;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Context;
import org.zeromq.ZMQ.PollItem;
import org.zeromq.ZMQ.Poller;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZMsg;

import com.google.gson.Gson;

public class ClientNode {

	public static void main(String[] args) throws IOException {

		Context context = ZMQ.context(1);
		Socket pubSocket = context.socket(ZMQ.PUB);
		Socket ackSocket = context.socket(ZMQ.ROUTER);
		Socket replySocket = context.socket(ZMQ.ROUTER);

		pubSocket.bind("tcp://*:5000");
		ackSocket.bind("tcp://*:5001");
		replySocket.bind("tcp://*:5002");

		ackSocket.setReceiveTimeOut(20);
		replySocket.setReceiveTimeOut(5 * 1000);

		while (!Thread.currentThread().isInterrupted()) {

			BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
			System.out.print("enter your search term:");
			String q = br.readLine();

			System.out.print("enter the categories you'd like to search (comma separated):");
			String[] categories = br.readLine().split(",");

			Search search = new Search();
			search.q = q;
			search.categories = categories;

			Gson gson = new Gson();

			String searchId = UUID.randomUUID().toString();

			System.out.println("search id: " + searchId);

			pubSocket.send("search_notification", ZMQ.SNDMORE);
			pubSocket.send(searchId, ZMQ.SNDMORE);
			pubSocket.send("tcp://localhost:5001", ZMQ.SNDMORE);
			pubSocket.send("tcp://localhost:5002", ZMQ.SNDMORE);
			pubSocket.send("UTF-8", ZMQ.SNDMORE);
			pubSocket.send(gson.toJson(search));

			ArrayList<ZMsg> acks = new ArrayList<>();

			Poller poller = new Poller(1);
			
			poller.register(ackSocket, Poller.POLLIN);

			while (true) {
			
				System.out.println(poller.poll(20));
				
				if (poller.pollin(0)) {
					System.out.println("-----------------------");
					String ackSearchId = ackSocket.recvStr();
					String nodeId = ackSocket.recvStr();

					System.out.println(ackSearchId);
					System.out.println(nodeId);
					System.out.println("-----------------------");
				} else {
					break;
				}
			}
		}
	}
}
